interfaces/policy/policy.go

```diff
@@ -0,0 +1,111 @@
+
+// ConnectCandidate represents a candidate connection.
+type ConnectCandidate struct {
+	// TODO: later we need to carry dynamic attributes once we have those
+	Plug                *snap.PlugInfo
+	PlugSnapDeclaration *asserts.SnapDeclaration
+
+	Slot                *snap.SlotInfo
+	SlotSnapDeclaration *asserts.SnapDeclaration
+
+	BaseDeclaration *asserts.BaseDeclaration
+}
+
+func (connc *ConnectCandidate) plugAttrs() map[string]interface{} {
+	return connc.Plug.Attrs
+}
+
+func (connc *ConnectCandidate) slotAttrs() map[string]interface{} {
+	return connc.Slot.Attrs
+}
+
+func (connc *ConnectCandidate) plugSnapType() snap.Type {
+	return connc.Plug.Snap.Type
+}
+
+func (connc *ConnectCandidate) slotSnapType() snap.Type {
+	return connc.Slot.Snap.Type
+}
+
+func (connc *ConnectCandidate) checkPlugRule(rule *asserts.PlugRule, whichDecl string) error {
+	if checkPlugConnectionConstraints(connc, rule.DenyConnection) == nil {
+		return fmt.Errorf("connection denied because it matches deny-connection in plug rule for interface %q from %s", connc.Plug.Interface, whichDecl)
+	}
+	if checkPlugConnectionConstraints(connc, rule.AllowConnection) != nil {
+		return fmt.Errorf("connection denied because it does not match allow-connection in plug rule for interface %q from %s", connc.Plug.Interface, whichDecl)
+	}
+	return nil
+}
+
+func (connc *ConnectCandidate) checkSlotRule(rule *asserts.SlotRule, whichDecl string) error {
+	if checkSlotConnectionConstraints(connc, rule.DenyConnection) == nil {
+		return fmt.Errorf("connection denied because it matches deny-connection in slot rule for interface %q from %s", connc.Plug.Interface, whichDecl)
+	}
+	if checkSlotConnectionConstraints(connc, rule.AllowConnection) != nil {
+		return fmt.Errorf("connection denied because it does not match allow-connection in slot rule for interface %q from %s", connc.Plug.Interface, whichDecl)
+	}
+	return nil
+}
+
+// Check checks whether the connection is allowed.
+func (connc *ConnectCandidate) Check() error {
+	baseDecl := connc.BaseDeclaration
+	if baseDecl == nil {
+		return fmt.Errorf("internal error: improperly initialized ConnectCandidate")
+	}
+	iface := connc.Plug.Interface
+
+	if plugDecl := connc.PlugSnapDeclaration; plugDecl != nil {
+		which := fmt.Sprintf("snap-declaration for snap %q (id %s)", plugDecl.SnapName(), plugDecl.SnapID())
+		if rule := plugDecl.PlugRule(iface); rule != nil {
+			return connc.checkPlugRule(rule, which)
+		}
+	}
+	if slotDecl := connc.SlotSnapDeclaration; slotDecl != nil {
+		which := fmt.Sprintf("snap-declaration for snap %q (id %s)", slotDecl.SnapName(), slotDecl.SnapID())
+		if rule := slotDecl.SlotRule(iface); rule != nil {
+			return connc.checkSlotRule(rule, which)
+		}
+	}
+	if rule := baseDecl.PlugRule(iface); rule != nil {
+		return connc.checkPlugRule(rule, "base-declaration")
+	}
+	if rule := baseDecl.SlotRule(iface); rule != nil {
+		return connc.checkSlotRule(rule, "base-declaration")
+	}
+	return nil
+}
```

Comment:
This matches my expectations for precedence: snap plug, snap slot, base plug, base slot, allow.